fixing bugs. Lets do this

gotibhai opened the pull request:
New things. 
1. Added Screen for username. Commented out for testing. 
2. Made the buttons flip color and the selected one is the dark one. 
3. Added quote at the bottom. 
4. Cleaned up timer code, Sorry I had to take out the timer.ref obj. We should only keep roundState in parent. I've also reduced from 3 states to 2. Running and Finished. But person can join anytime. This is tested and works. 
5. When we're waiting on new round to start, Instead of showing empty graph, I generate a sin graph just to make it look cool!

**Main Problem: f(x) is somehow only producing once? Or node isn't able to fetch the second round. So 2nd round is janky.** 

Edit: This is transient?? It worked the last couple of times. Please make sure there's no race condition or something.

Todo: 
[ ] Consume predictions and rank the usernames with a ranking function. 
[ ] Add pipelining for results and send it to front end. Graph should work in some basic form.
[ ] Remaining stuff on Notion
[ ] DW about graphs too much, I'll look into a better library when we're polishing.

abdulmi commented:
1. Added ranking of users submissions
2. Added sending users rankings to the results_stream & listening on the results stream on Node

abdulmi commented:
Todo:
1. We should enable users so we can see each user's submission
2. View the rankings in a nice way(table, or graph,..etc)

gotibhai commented on server/main.py:
```diff
@@ -8,22 +8,57 @@
+    for i in range(len(users_predictions)):
+        users_predictions[i] = users_predictions[i].decode('utf-8')
+    # print("before sorting ", users_predictions)
+    users_predictions = sorted(users_predictions, key=lambda p: abs(int(p.split(":")[1])-delta))
```
@abdulmi I'd prefer if we gave them scores which were dependant on their location in the queue and their closeness to the final value. This will be more interesting to see. You don't need to sort we can just sort on frontend.

gotibhai replied:
This is needed for V2 please, it'll be godly to have scores rather than ranking.

abdulmi replied:
yea that was my next plan of attack; working on it!

gotibhai commented on server/main.py:
```diff
@@ -3,35 +3,40 @@
+redis_client = redis.Redis(host='redis_db', port=6379, db=0)
+
+TIME_TILL_ROUND_START = 10
+ROUND_TIME = 10
```
Let's use constants please easier to manipulate.

gotibhai commented on server/main.py:
```diff
@@ -3,35 +3,40 @@
-        r.xadd("results_stream", {"rankings": rankings})
+        users_rankings = get_users_rankings(new_values)
+        #redis_client.xadd("results_stream", {"rankings": rankings})
+        redis_client.publish("round_result", json.dumps(users_rankings))
```
Changing round result from streams -> pubsub. Much faster, streams was delayed and didn't work as fast.

abdulmi replied:
nice one!

gotibhai commented on server/main.py:
```diff
@@ -101,4 +106,5 @@ def f_x():
 
 
 if __name__ == '__main__':
+    redis_client.delete("user_prediction")
```
Did this to remove old janky data just in case. Can remove it / or not doesn't matter.

gotibhai commented:
Added Quite a few UI changes, fixed graphs, fixed timer and other coordination issues. Made it work end to end. 

@abdulmi
1. Could you please look into integrating with something like this https://finnhub.io/docs/api#stock-candles to pull live stock data?
2. Make a ranking function which gives scores please? (Stretch goal: Can we investigate into using Redis Leaderboard for this. That'll be a huge win for us)
3. Start diving into other small fixes we mentioned?
